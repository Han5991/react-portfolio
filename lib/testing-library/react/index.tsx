/* eslint-disable import/no-extraneous-dependencies */
import {render, RenderOptions} from '@testing-library/react';
import React, {ReactElement} from 'react';

import {ThemeProvider} from '@lib/styled-components';
import theme from '@styles/theme';

const Providers = ({children}: {children: React.ReactNode}) => (
  <ThemeProvider theme={{...theme.light}}>{children}</ThemeProvider>
);

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, {wrapper: Providers, ...options});

export * from '@testing-library/react';
export {customRender as render};
