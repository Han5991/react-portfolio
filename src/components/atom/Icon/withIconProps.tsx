import React, {FC, SVGProps} from 'react';

import {useTheme} from '@lib/styled-components';

export type IconProps = SVGProps<SVGSVGElement>;

const withIconProps =
  (Icon: FC<IconProps>) =>
  // eslint-disable-next-line react/display-name
  (iconProps: IconProps) => {
    const theme = useTheme();
    const defaultFill = theme.color.text[700];
    const defaultSize = theme.size[6];
    const {fill, width, height, ...restProps} = iconProps;
    return (
      <Icon
        data-testid="@icon"
        fill={fill ?? defaultFill}
        width={width ?? defaultSize}
        height={height ?? defaultSize}
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...restProps}
      />
    );
  };

export default withIconProps;
