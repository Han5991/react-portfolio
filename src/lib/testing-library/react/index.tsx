/* eslint-disable import/no-extraneous-dependencies */
import {render, RenderOptions} from '@testing-library/react';
import React, {ReactElement, ReactNode} from 'react';

import {RecoilRoot} from '@lib/recoil';
import {ThemeProvider} from '@lib/styled-components';
import theme from '@styles/theme';

const matchMediaMock = jest.fn().mockReturnValue({
  matches: true,
  addEventListener: jest.fn(),
  removeEventListener: jest.fn(),
});

beforeAll(() => {
  window.matchMedia = matchMediaMock;
});

const Providers = ({children}: {children: ReactNode}) => (
  <RecoilRoot>
    <ThemeProvider theme={theme.light}>{children}</ThemeProvider>
  </RecoilRoot>
);

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>,
) => render(ui, {wrapper: Providers, ...options});

export * from '@testing-library/react';
export {customRender as render};
