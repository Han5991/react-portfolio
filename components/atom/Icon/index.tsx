import {FC, SVGProps} from 'react';

import withIconProps from './withIconProps';

import * as svgs from '@assets/svgs';

export type SvgName = keyof typeof svgs;
const svgNameList = Object.keys(svgs) as SvgName[];

const SvgMap = {} as {
  [k in SvgName]: FC<SVGProps<SVGSVGElement>>;
};

svgNameList.forEach((svgName: SvgName) => {
  SvgMap[svgName] = withIconProps(svgs[svgName]);
});

export default SvgMap;
