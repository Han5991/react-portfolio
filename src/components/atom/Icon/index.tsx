import {FC} from 'react';

import withIconProps, {IconProps} from './withIconProps';

import * as svgs from '@assets/svgs';

export type SvgName = keyof typeof svgs;
const svgNameList = Object.keys(svgs) as SvgName[];

const SvgMap = {} as {
  [k in SvgName]: FC<IconProps>;
};

svgNameList.forEach((svgName: SvgName) => {
  SvgMap[svgName] = withIconProps(svgs[svgName]);
});

export default SvgMap;
