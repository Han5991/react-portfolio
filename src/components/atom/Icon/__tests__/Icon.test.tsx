import React from 'react';

import Icon from '../index';

import {render, screen} from '@lib/testing-library/react';

describe('Icon component test', () => {
  test('should render default props snapshot', () => {
    const {container} = render(<Icon.HamburgerButton />);
    expect(container).toMatchSnapshot();
  });

  test('should render custom props snapshot', () => {
    const fill = 'red';
    const width = 100;
    const height = 200;
    render(<Icon.HamburgerButton fill={fill} width={width} height={height} />);
    const svgElement = screen.getByTestId('@icon');
    expect(svgElement).toBeInTheDocument();
    expect(svgElement).toHaveAttribute('fill', fill);
    expect(svgElement).toHaveAttribute('width', `${width}`);
    expect(svgElement).toHaveAttribute('height', `${height}`);
  });
});
