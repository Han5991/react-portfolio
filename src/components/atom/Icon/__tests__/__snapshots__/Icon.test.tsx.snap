// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon component test should render default props snapshot 1`] = `
<div>
  <svg
    data-testid="@icon"
    fill="#282C34"
    height="24"
    width="24"
  />
</div>
`;
